File: src/main/java/com/letgo/book/infrastructure/persistence/mapping/HibernateBookEntity.java
package com.letgo.book.infrastructure.persistence.mapping;

import com.letgo.book.domain.BookId;
import com.letgo.book.domain.BookTitle;
import org.hibernate.annotations.Type;

import javax.persistence.*;

@Entity
@Table(name = "books")
final public class HibernateBookEntity {
    @Id
    @Type(type = "com.letgo.book.infrastructure.persistence.mapping.BookIdType")
    private BookId id;
    @Convert(converter = BookTitleConverter.class)
    @Column(nullable = false)
    private BookTitle title;

    public HibernateBookEntity() {
    }

    public HibernateBookEntity(BookId id, BookTitle title) {
        this.id = id;
        this.title = title;
    }

    public BookId id() {
        return id;
    }

    public BookTitle title() {
        return title;
    }
}

File: src/main/java/com/letgo/book/infrastructure/persistence/mapping/HibernateBookEntity.kt
package com.letgo.book.infrastructure.persistence.mapping

import com.letgo.book.domain.BookId
import com.letgo.book.domain.BookStatus
import org.hibernate.annotations.Type
import javax.persistence.*

@Entity
@Table(name = "books")
class HibernateBookEntity {
    @Id
    @Type(type = "com.letgo.book.infrastructure.persistence.mapping.BookIdType")
    private var id: BookId? = null

    @Embedded
    private var title: BookTitleEmbeddable? = null

    @Enumerated(EnumType.ORDINAL)
    private var status: BookStatus? = null

    constructor()
    constructor(id: BookId, title: BookTitleEmbeddable, status: BookStatus) {
        this.id = id
        this.title = title
        this.status = status
    }

    fun id(): BookId? {
        return id
    }

    fun title(): BookTitleEmbeddable? {
        return title
    }

    fun status(): BookStatus? {
        return status
    }
}
